### PR TITLE
asadiqbal08/ENT-1232 Unenrollments in Learner Reports 

### PR DESCRIPTION
**Pull request by asadiqbal08**
Unenrollments in Learner Reports [ENT-1232](https://openedx.atlassian.net/browse/ENT-1232)

As an Enterprise customer, I want to be able to have my employees unenroll in a course within the 14 day refund period so they have the opportunity to see if the course is or is not the right one for their needs.
As an edX operator, when a learner unenrolls, I want to know the enrollment date, and whether it occurred within the refund period

### CODE REVIEW

File: enterprise_data/api/v0/serializers.py [hunk 1]
```diff
@@ -14,13 +14,30 @@ class EnterpriseEnrollmentSerializer(serializers.ModelSerializer):
...
+        """
+        if obj.unenrollment_timestamp:
+            period = range(0, 15)
+            unenrollment_within_date = (obj.unenrollment_timestamp - obj.course_start).days in period or \
+                                       (obj.unenrollment_timestamp - obj.enrollment_created_timestamp).days in period
+
+            if unenrollment_within_date:
+                return True
+
+        return False
```
Comment (brittneyexline) [inline enterprise_data/api/v0/serializers.py]: i think it would be better to return None when there is no unenrollment to avoid confusion.

File: enterprise_data/api/v0/serializers.py [hunk 2]
```diff
@@ -14,13 +14,30 @@ class EnterpriseEnrollmentSerializer(serializers.ModelSerializer):
...
         )
 
+    def get_unenrollment_end_within_date(self, obj):
+        """
+        Return "True" when un-enrolled date is within 14 days of course start or enrollment date
+        (whichever is later), when un-enrolled date is after 14 days of
+        course start or enrollment date (whichever is later) return "False".
+        """
+        if obj.unenrollment_timestamp:
+            period = range(0, 15)
```
Comment (brittneyexline) [inline enterprise_data/api/v0/serializers.py]: it might be good to define this as a constant, in case we change our policy.

File: enterprise_data/api/v0/serializers.py [hunk 3]
```diff
@@ -14,13 +14,30 @@ class EnterpriseEnrollmentSerializer(serializers.ModelSerializer):
...
+            unenrollment_within_date = (obj.unenrollment_timestamp - obj.course_start).days in period or \
```
Comment (brittneyexline) [inline enterprise_data/api/v0/serializers.py]: nit: we could just use simple comparisons (>0, <=14) instead of checking if an integer is in a list. it would be slightly more performant and readable imo.

**Review by brittneyexline — COMMENTED** [review]
just a few small things to address, plus adding tests!

**Comment by asadiqbal08** [comment]
@brittneyexline kindly take another look.

File: enterprise_data/api/v0/serializers.py [hunk 4]
```diff
@@ -14,13 +14,29 @@ class EnterpriseEnrollmentSerializer(serializers.ModelSerializer):
...
         )
 
+    def get_unenrollment_end_within_date(self, obj):
+        """
+        Return "True" when un-enrolled date is within 14 days of course start or enrollment date
+        (whichever is later), when un-enrolled date is after 14 days of
+        course start or enrollment date (whichever is later) return "None".
+        """
+        if obj.unenrollment_timestamp:
+            unenrollment_within_date = (0 < (obj.unenrollment_timestamp - obj.course_start).days <= 14) or \
```
Comment (brittneyexline) [inline enterprise_data/api/v0/serializers.py:32]: shouldn't we just return whatever this evaluates to? as is, we would return None when we probably should return False

**Review by brittneyexline — COMMENTED** [review]
i think adding test cases will be good to clarify the logic you have in the serializer. also keep in mind that this should not be merged until you analytics pipeline PR has merged and run in production.

File: enterprise_data/api/v0/serializers.py (hunk 4 above)
Comment (asadiqbal08) [reply enterprise_data/api/v0/serializers.py:32]: I prefer to return either `True` or `False` instead of `None`. I agree, we just return whatever this evaluates to `True/False`

File: enterprise_reporting/export_edxapp_data.py [hunk 5]
```diff
@@ -34,9 +34,10 @@
 import sys
 from io import open  # pylint: disable=redefined-builtin
 
-import mysql.connector
 import vertica_python
 
+import mysql.connector
```
Comment (zubair-arbi) [inline enterprise_reporting/export_edxapp_data.py]: Can you please revert this files change? as it seems irrelevant.

**Review by zubair-arbi — APPROVED** [review]
Just one small suggestion otherwise LGTM 👍

File: enterprise_reporting/export_edxapp_data.py (hunk 5 above)
Comment (asadiqbal08) [reply enterprise_reporting/export_edxapp_data.py]: The change here is showing by `isort` command. That import is in use.

**Comment by asadiqbal08** [comment]
@brittneyexline your final review in it ...

File: enterprise_data/api/v0/serializers.py [hunk 6]
```diff
@@ -14,13 +14,27 @@ class EnterpriseEnrollmentSerializer(serializers.ModelSerializer):
...
             enterprise_id=obj.enterprise_id, course_id=obj.course_id
         )
 
+    def get_unenrollment_end_within_date(self, obj):
+        """
+        Return "True" when un-enrolled date is within 14 days of course start or enrollment date
+        (whichever is later), when un-enrolled date is after 14 days of
+        course start or enrollment date (whichever is later) return "False".
+        """
+        unenrollment_within_date = False
```
Comment (brittneyexline) [inline enterprise_data/api/v0/serializers.py]: i think this should default to None if there is no unenrollment timestamp. i know you disagreed earlier, but to me, it is misleading for it to be false when there is no unenrollment timestamp to evaluate. why not make each case return a distinct value instead of overloading false?

File: enterprise_data/fixtures/enterprise_enrollment.json [hunk 7]
```diff
@@ -34,7 +34,8 @@
       "current_grade": 0.80,
       "course_price": 200.00,
       "discount_price": 120.00,
-      "created": "2018-08-01 23:14:35Z"
+      "created": "2018-08-01 23:14:35Z",
+      "unenrollment_timestamp": "2014-06-29 16:02:38Z"
```
Comment (brittneyexline) [inline enterprise_data/fixtures/enterprise_enrollment.json]: setting unenrollment timestamp for all the fixture data means we are missing the case where the user has not unenrolled, which is likely to be the most common case. you should leave at least one of these rows with unenrollment timestamp being null.

**Comment by asadiqbal08** [comment]
@brittneyexline Updated

**Review by brittneyexline — APPROVED** [review]

**Comment by brianhw** [comment]
Is this fixture supposed to exactly match the table that is generated by the pipeline in the resultstore?   This PR sets up the table to have the unenrollment_timestamp field appear after the created field, and that's not the case.   The 'created' field is always last.

**Comment by asadiqbal08** [comment]
@brianhw this [fixture](https://github.com/edx/edx-enterprise-data/blob/master/enterprise_data/fixtures/enterprise_enrollment.json) is used in development process for testing. So what do you suggest, should I create a new PR with re-order the columns structure as you mentioned the "created" field is always be the last in order  or should do in separate ticket ?
